backends/ref/ceed-ref-basis.c

```diff
@@ -191,36 +191,31 @@ static int CeedBasisApply_Ref(CeedBasis basis, CeedInt num_elem, CeedTransposeMo
     }
   } else {
     // Non-tensor basis
+    CeedInt P = num_nodes, Q = num_qpts;
```

Comment:
I'm curious about this change: how was this able to be pulled out?  Previously, Q was defined differently for `CEED_EVAL_INTERP` (Q_comp * num_qpts).  Also, how do you avoid needing to swap P and Q based on `t_mode`?

Reply:
Hi @nbeams, if you look at the new `CeedTensorContractStridedApply` function, this handles both of the comments you have (swaps `P` and `Q` for `t_mode` and also handles `q_comp`). This was introduced to avoid duplicating the same code for all of the `eval_mode` cases.

Reply:
Oh, ha, I didn't even notice the name change 🤦‍♀️ .  But I'm still not sure how `INTERP` and `GRAD` could have the exact same arguments (except for the basis matrix `interp` or `grad`) given to the new function.  How does it know that the gradient matrix is of size `P*dim x Q` while `interp` is `P x Q`?

Reply:
For H^1 `INTERP`, `q_comp = 1` versus for `GRAD`, `q_comp = dim`. This for H^1 `INTERP` the "strided apply" doesn't really do anything over the `q_comp` direction. The idea here was to just generalize to a basis tensor of shape `q_comp x num_qpts x num_nodes` to capture both the H^1 `INTERP` and `GRAD` operations, as well as the various shapes of H(div) and H(curl) basis operations.

Reply:
Oh, I see, `q_comp` is no longer a property of the basis object.  I'm not sure how I feel about that change, but if other developers like it, then I won't object. 
The cuda-ref and hip-ref (and MAGMA) non-tensor basis implementations don't use the CeedTensorContractApply setup, but have specialized kernels, so I'm not sure how this refactoring will really help them, but it obviously doesn't hurt anything to have it in ref.